src/rules/newlineRule.ts

```diff
@@ -0,0 +1,150 @@
+            * \`"class"\` checks for empty line after class declaration.
+            * \`"functionBlock"\` checks for empty line after function block declaration.
+            * \`"block"\` checks for empty line after block declaration.`,
+        options: {
+            properties: {
+                return: ALWAYS_IGNORE_OR_NEVER,
+                class: ALWAYS_IGNORE_OR_NEVER,
+                functionBlock: ALWAYS_IGNORE_OR_NEVER,
+                block: ALWAYS_IGNORE_OR_NEVER,
+            },
+            type: "object",
+        },
+        optionExamples: [[true, {return: "always", class: "never"}]],
+        type: "style",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+ 
+    public static FAILURE_STRING_FACTORY(kind: boolean, nodeType: string) {
+        const kindMsg = kind ? 'Missing' : 'Unneeded';
+        return `${kindMsg} blank line before ${nodeType}`;
+    }
+ 
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NewlineWalker(sourceFile, this.ruleName, parseOptions(this.ruleArguments[0] as Option | Options | undefined)));
+    }
+}
+ 
+class NewlineWalker extends Lint.AbstractWalker<Options> {
+    public walk(sourceFile: ts.SourceFile) {
+        const cb = (node: ts.Node): void => {
+            if (node.kind === ts.SyntaxKind.ReturnStatement && this.options.return != OptionState.ignore) {
+                const prev = getPreviousStatement(node as ts.Statement);
+                if (prev === undefined) {
+                    // return is not within a block (e.g. the only child of an IfStatement) or the first statement of the block
+                    // no need to check for preceding newline
+                    return;
+                }
+ 
+                this.checkForEmptyLine(prev, node, this.options.return == OptionState.always, 'return');
+            }
+ 
+            if (isClassDeclaration(node) && this.options.class != OptionState.ignore) {
+                this.checkForEmptyLine(
+                    getChildOfKind(node, ts.SyntaxKind.OpenBraceToken, this.sourceFile)!, node.members[0],
+                    this.options.class == OptionState.always, 'first member',
+                );
+            }
+ 
+            if (isBlock(node) &&
+                    (isFunctionWithBody(node.parent!) && this.options.functionBlock != OptionState.ignore
+                        || this.options.block != OptionState.ignore)) {
+                this.checkForEmptyLine(
+                    node.getChildAt(0), node.statements[0], this.options.functionBlock == OptionState.always, 'first statement',
+                );
+            }
+            return ts.forEachChild(node, cb);
+        };
+        return ts.forEachChild(sourceFile, cb);
+    }
+ 
+    private checkForEmptyLine(node: ts.Node, firstNode: ts.Node, lineRequired: boolean, nodeType: string) {
+        let start = node.end;
```

Comment:
consider replacing the `node: ts.Node` parameter with `start: number` and pass the position directly. That avoids accessing tokens with `getChildAt`